File: solutions/javascript/queue.js
var Queue = module.exports = function () {
  this.head   = null;
  this.tail   = null;
  this.length = 0;
};

Queue.prototype.enqueue = function (value) {
  var node = {
    value: value,
    next: null
  };

  // If there is currently no head node, set it to the current node.
  if (!this.head) {
    this.head = node;
  }

  // If we have a tail node already, set it's next property to be the current
  // node.
  if (this.tail) {
    this.tail.next = node;
  }

  // Update the tail to be the next node.
  this.tail = node;

  return this.length += 1;
};

Queue.prototype.dequeue = function () {
  if (!this.head) { return; }

  var node = this.head;

  // Update the head reference and remove the next node reference from the
  // previous head.
  this.head = node.next;
  node.next = null;

  // Remove the tail node if we have no more head node.
  if (!this.head) { this.tail = null; }

  this.length -= 1;

  return node.value;
};

File: solutions/java/Queue.java
// simple implementation of Queue with linked list
public class Queue {

	private Node first, last;


	private class Node{
		private Node next;
		private int data;

		Node(int data, Node next) {
			this.data = data;
			this.next = next;
		}
	}

	public void enqueue(int data) {
		Node n = new Node(data, null);
		if (first == null) {
			first = n;
			last = n;
			return;
		}

		last.next = n;
		last = n;
	}

	public int dequeue() throws Exception {
		if (first == null) {
			throw new Exception("Queue is empty, nothing to dequeue.");
		}
		int store = first.data;
		first = first.next;

		return store;
	}

	public void printQueue() {
		Node current = first;
		while (current != null) {
			System.out.println(current.data);
			current = current.next;
		}
	}
	public static void main(String[] args) throws Exception {
		Queue queue = new Queue();

		queue.enqueue(1);
		queue.enqueue(2);
		queue.enqueue(3);
		queue.enqueue(4);
		queue.enqueue(5);
		queue.enqueue(6);

		queue.printQueue();

		queue.dequeue();
		queue.dequeue();
		queue.dequeue();
		queue.dequeue();
		queue.printQueue();

	}
}